scvi/autotune/_autotune.py

```diff
@@ -0,0 +1,140 @@
+
+
+class Autotune:
+    """
+
+    Hyperparameter tuning for SCVI using Ray Tune.
+
+    Parameters
+    ----------
+    adata
+        AnnData object we will tune the model on.
+    model
+        Model from scvi.model we will tune.
+    training_metrics
+        Metrics to track during training.
+    metric_functions
+        For metrics calculated after training a model, like silhouette distance.
+    model_hyperparams
+        Config for the model hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    trainer_hyperparams
+        Config for the trainer hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    plan_hyperparams
+        Config for the training_plan hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    """
+
+    def __init__(
+        self,
+        adata,
+        model,
+        training_metrics: list = ["elbo_validation"],
+        metric_functions: dict = {},
+        model_hyperparams: dict = {},
+        trainer_hyperparams: dict = {},
+        plan_hyperparams: dict = {},
+        num_epochs: int = 2,
+    ):
+        self.adata = adata
+        self.model = model
+        self.training_metrics = training_metrics
+        self.metric_functions = metric_functions
+        self.model_hyperparams = model_hyperparams
+        self.trainer_hyperparams = trainer_hyperparams
+        self.plan_hyperparams = plan_hyperparams
+        self.metrics = training_metrics
+        self.reporter = CLIReporter(
+            metric_columns=training_metrics + list(self.metric_functions.keys())
+        )
+        self.config = {}
+        for d in [model_hyperparams, trainer_hyperparams, plan_hyperparams]:
+            if d is not None:
+                self.config.update(d)
+        self.num_epochs = num_epochs
+
+    def _scvi_trainable(self, config, checkpoint_dir=None):
+        model_config = {}
+        trainer_config = {}
+        plan_config = {}
+        for key in config:
+            if key in self.model_hyperparams:
+                model_config[key] = config[key]
+            elif key in self.trainer_hyperparams:
+                trainer_config[key] = config[key]
+            elif key in self.plan_hyperparams:
+                plan_config[key] = config[key]
+
+        _model = self.model(self.adata, **model_config)
+        _model.train(
+            **trainer_config,
+            plan_kwargs=plan_config,
+            callbacks=[
+                ModelSave(_model),
+                _TuneReportMetricFunctionsCallback(
+                    metrics=self.metrics,
+                    on="validation_end",
+                    model=_model,
+                    metric_functions=self.metric_functions,
+                ),
+            ],
+            check_val_every_n_epoch=1,
+            max_epochs=self.num_epochs,
+        )
+
+    def run(
+        self,
+        metric,
+        scheduler,
```

Comment:
should there be a default scheduler? So default is None and if None, use Asha...